app/src/main/java/com/techcourse/dao/UserDao.java

```diff
@@ -1,116 +1,92 @@
 package com.techcourse.dao;
 
 import com.techcourse.domain.User;
+import nextstep.jdbc.JdbcTemplate;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import javax.sql.DataSource;
-import java.sql.Connection;
-import java.sql.PreparedStatement;
-import java.sql.ResultSet;
-import java.sql.SQLException;
 import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 public class UserDao {
 
     private static final Logger log = LoggerFactory.getLogger(UserDao.class);
 
-    private final DataSource dataSource;
+    private JdbcTemplate jdbcTemplate;
 
     public UserDao(DataSource dataSource) {
-        this.dataSource = dataSource;
+        this.jdbcTemplate = new JdbcTemplate(dataSource);
     }
 
     public void insert(User user) {
         final String sql = "insert into users (account, password, email) values (?, ?, ?)";
-
-        Connection conn = null;
-        PreparedStatement pstmt = null;
-        try {
-            conn = dataSource.getConnection();
-            pstmt = conn.prepareStatement(sql);
-
-            log.debug("query : {}", sql);
-
-            pstmt.setString(1, user.getAccount());
-            pstmt.setString(2, user.getPassword());
-            pstmt.setString(3, user.getEmail());
-            pstmt.executeUpdate();
-        } catch (SQLException e) {
-            log.error(e.getMessage(), e);
-            throw new RuntimeException(e);
-        } finally {
-            try {
-                if (pstmt != null) {
-                    pstmt.close();
-                }
-            } catch (SQLException ignored) {}
-
-            try {
-                if (conn != null) {
-                    conn.close();
-                }
-            } catch (SQLException ignored) {}
-        }
+        log.debug("query : {}", sql);
+        jdbcTemplate.queryDML(sql, user.getAccount(), user.getPassword(), user.getEmail());
     }
 
     public void update(User user) {
-        // todo
+        final String sql = "update users set account=?, password=?, email=? where id=?";
+        log.debug("query : {}", sql);
+        jdbcTemplate.queryDML(sql, user.getAccount(), user.getPassword(), user.getEmail(), user.getId());
     }
 
     public List<User> findAll() {
-        // todo
-        return null;
+        final String sql = "select id as users_id, account as users_account, " +
+                "password as users_password, email as users_email " +
+                "from users";
+        log.debug("query : {}", sql);
+        final List<Map<String, Object>> result = jdbcTemplate.queryForList(sql);
```

Comment:
DB 결과를 Map으로 반환하면 라이브러리 사용자가 사용하기 번거롭습니다.
Map이 어떤 구조로 이루어져 있는지 파악해야되고 그에 맞춰 또 매핑이 필요합니다.
어떻게 하면 사용자가 편하게 작업할 수 있을지 고민해보는게 좋겠네요.

Reply:
구구 의견에 동의합니다. datasource에서 jdbcTemplate에게 객체 매핑 (이 상황에선 User)에 관련된 정보를 준다면 jdbcTemplate이 뭔가 더 해줄 수 있을 것 같네요~
그러면 지금처럼 Map을 가져와서 datasource에서 매핑하지않고 바로 매핑된 객체를 받을 수 있지 않을까요? 아래는 이해를 돕기위한 참고용입니다!

```suggestion
        final List<User> result = jdbcTemplate.queryForList(sql, 매핑과 관련된 정보);
```

Reply:
흠.. query라는 메소드가 이미 그 기능을 하고 있어, 상관없다고 생각했었어요.

```java
    public <T> List<T> query(String sql, @Nullable Object[] args, RowMapper<T> rowMapper) {
        return query(sql, args, new RowMapperResultSetExtractor<>(rowMapper));
    }
```

테스트 코드가 없어, 의도를 정확하게 드러내지 못한 것 같아 테스트 코드또한 추가를 완료했어요! 
단순히 작성만 하면, 상대 개발자가 의도를 파악하기 힘들다는 것을 알게됐어요! 


b08ba6e